java/vector/src/main/java/org/apache/arrow/vector/complex/LargeListVector.java

```diff
@@ -0,0 +1,991 @@
+ * </ol>
+ *
+ * This is the LargeList variant of list, it has a 64-bit wide offset
+ *
+ * <p>
+ *   todo review checkedCastToInt usage in this class.
+ *   Once int64 indexed vectors are supported these checks aren't needed.
+ * </p>
+ */
+public class LargeListVector extends BaseValueVector implements RepeatedValueVector, BaseListVector, PromotableVector {
+
+  public static LargeListVector empty(String name, BufferAllocator allocator) {
+    return new LargeListVector(name, allocator, FieldType.nullable(ArrowType.LargeList.INSTANCE), null);
+  }
+
+  public static final FieldVector DEFAULT_DATA_VECTOR = ZeroVector.INSTANCE;
+  public static final String DATA_VECTOR_NAME = "$data$";
+
+  public static final byte OFFSET_WIDTH = 8;
+  protected ArrowBuf offsetBuffer;
+  protected FieldVector vector;
+  protected final CallBack callBack;
+  protected int valueCount;
+  protected long offsetAllocationSizeInBytes = INITIAL_VALUE_ALLOCATION * OFFSET_WIDTH;
+  private final String name;
+
+  protected String defaultDataVectorName = DATA_VECTOR_NAME;
+  protected ArrowBuf validityBuffer;
+  protected UnionLargeListReader reader;
+  private final FieldType fieldType;
+  private int validityAllocationSizeInBytes;
+
+  /**
+   * The maximum index that is actually set.
+   */
+  private long lastSet;
+
+  /**
+   * Constructs a new instance.
+   *
+   * @param name The name of the instance.
+   * @param allocator The allocator to use for allocating/reallocating buffers.
+   * @param fieldType The type of this list.
+   * @param callBack A schema change callback.
+   */
+  public LargeListVector(String name, BufferAllocator allocator, FieldType fieldType, CallBack callBack) {
+    super(allocator);
+    this.name = name;
+    this.validityBuffer = allocator.getEmpty();
+    this.fieldType = checkNotNull(fieldType);
+    this.callBack = callBack;
+    this.validityAllocationSizeInBytes = getValidityBufferSizeFromCount(INITIAL_VALUE_ALLOCATION);
+    this.lastSet = -1;
+    this.offsetBuffer = allocator.getEmpty();
+    this.vector = vector == null ? DEFAULT_DATA_VECTOR : vector;
+    this.valueCount = 0;
+  }
+
+  @Override
+  public void initializeChildrenFromFields(List<Field> children) {
+    if (children.size() != 1) {
+      throw new IllegalArgumentException("Lists have only one child. Found: " + children);
+    }
+    Field field = children.get(0);
+    AddOrGetResult<FieldVector> addOrGetVector = addOrGetVector(field.getFieldType());
+    if (!addOrGetVector.isCreated()) {
+      throw new IllegalArgumentException("Child vector already existed: " + addOrGetVector.getVector());
+    }
+
+    addOrGetVector.getVector().initializeChildrenFromFields(field.getChildren());
+  }
+
+  @Override
+  public void setInitialCapacity(int numRecords) {
+    validityAllocationSizeInBytes = getValidityBufferSizeFromCount(numRecords);
+    offsetAllocationSizeInBytes = (numRecords + 1) * OFFSET_WIDTH;
```

Comment:
We need to cast (numRecords + 1) to long. Otherwise, integer overflow may happen before promoting the result to long.

Reply:
done